pkg/entities/record.go

```diff
@@ -99,200 +104,320 @@ func (d *dataRecord) PrepareRecord() (uint16, error) {
+func (d *dataRecord) GetDataElements() []*InfoElementValue {
+	return d.elements
 }
 
 func (t *templateRecord) PrepareRecord() (uint16, error) {
 	// Add Template Record Header
-	header := make([]byte, 4)
-	binary.BigEndian.PutUint16(header[0:2], t.templateID)
-	binary.BigEndian.PutUint16(header[2:4], t.fieldCount)
-
-	_, err := t.buff.Write(header)
+	initialLength := t.buff.Len()
+	err := util.Encode(&t.buff, binary.BigEndian, t.templateID, t.fieldCount)
 	if err != nil {
 		return 0, fmt.Errorf("AddInfoElement(templateRecord) error in writing template header: %v", err)
 	}
-
-	return uint16(len(header)), nil
+	return uint16(t.buff.Len() - initialLength), nil
 }
 
-func (t *templateRecord) AddInfoElement(element *InfoElement, val interface{}) (uint16, error) {
+func (t *templateRecord) AddInfoElement(element *InfoElement, val interface{}, isDecoding bool) (uint16, error) {
 	// val could be used to specify smaller length than default? For now assert it to be nil
 	if val != nil {
 		return 0, fmt.Errorf("AddInfoElement(templateRecord) cannot take value %v (nil is expected)", val)
 	}
-	// Add field specifier
-	fieldSpecifier := make([]byte, 4, 8)
-	binary.BigEndian.PutUint16(fieldSpecifier[0:2], element.ElementId)
-	binary.BigEndian.PutUint16(fieldSpecifier[2:4], element.Len)
+	initialLength := t.buff.Len()
+	// Add field specifier {elementID: uint16, elementLen: uint16}
+	err := util.Encode(&t.buff, binary.BigEndian, element.ElementId, element.Len)
+	if err != nil {
+		return 0, err
+	}
 	if element.EnterpriseId != 0 {
 		// Set the MSB of elementID to 1 as per RFC7011
-		fieldSpecifier[0] = fieldSpecifier[0] | 0x80
-		bytesToAppend := make([]byte, 4)
-		binary.BigEndian.PutUint32(bytesToAppend, element.EnterpriseId)
-		fieldSpecifier = append(fieldSpecifier, bytesToAppend...)
-	}
-
-	bytesWritten, err := t.buff.Write(fieldSpecifier)
-	if err != nil {
-		return 0, fmt.Errorf("AddInfoElement(templateRecord) error in writing to buffer: %v", err)
+		t.buff.Bytes()[initialLength] = t.buff.Bytes()[initialLength] | 0x80
+		err = util.Encode(&t.buff, binary.BigEndian, element.EnterpriseId)
+		if err != nil {
+			return 0, err
+		}
 	}
-	t.templateElements = append(t.templateElements, element)
+	t.elements = append(t.elements, element)
 	// Keep track of minimum data record length required for sanity check
 	if element.Len == VariableLength {
 		t.minDataRecLength = t.minDataRecLength + 1
 	} else {
 		t.minDataRecLength = t.minDataRecLength + element.Len
 	}
-	return uint16(bytesWritten), nil
+	return uint16(t.buff.Len() - initialLength), nil
 }
 
 func (t *templateRecord) GetTemplateElements() []*InfoElement {
-	return t.templateElements
+	return t.elements
 }
 
 func (t *templateRecord) GetMinDataRecordLen() uint16 {
 	return t.minDataRecLength
 }
+
+// convertToType is to decode or encode data to specific type
+func convertToType(dataType IEDataType, val interface{}, isDecoding bool) (interface{}, error) {
+	if isDecoding {
```

Comment:
can this if else block be removed if we use a variable for function pointer that is initialized to either util.Decode or util.Encode?

Reply:
The only concern is that for encoding, firstly we need to check whether the provided value is the format required data type. For decoding, we don't have to do the checking, that makes the code different for handling encoding/decoding.

Reply:
Also for decoding, an empty interface of corresponding type for storing the result needs to be passed to function.

Reply:
We can probably move the type check with switch functionality for encoding to a different function. 
Or if refactoring is difficult, maybe we should define separate functions for encoding and decoding to cover the functionality of convertToIEDataType. Functionality seems to be different as well.. for encoding we are writing data to msg buffer and for decoding we are writing to info elements.